git-annex-remote-rclone

```diff
@@ -16,268 +14,175 @@
-			case "$op" in
-				STORE)
-					# Store the file to a location
-					# based on the key.
-					# XXX when at all possible, send PROGRESS
-					calclocation "$key"
-					if [ ! -e "$file" ]; then
-						echo TRANSFER-FAILURE STORE "$key" "asked to store non-existent file $file"
-					else
-						if runcmd rclone copy "$file" "$LOC"; then
-							echo TRANSFER-SUCCESS STORE "$key"
-						else
-							echo TRANSFER-FAILURE STORE "$key"
-						fi
-					fi
-				;;
-				RETRIEVE)
-					# Retrieve from a location based on
-					# the key, outputting to the file.
-					# XXX when easy to do, send PROGRESS
-					calclocation "$key"
-					# http://stackoverflow.com/questions/31396985/why-is-mktemp-on-os-x-broken-with-a-command-that-worked-on-linux
-					if GA_RC_TEMP_DIR=$(mktemp -d "${TMPDIR:-/tmp}/rclone-annex-tmp.XXXXXXXXX") &&
-						runcmd rclone copy "$LOC$key" "$GA_RC_TEMP_DIR" &&
-						mv "$GA_RC_TEMP_DIR/$key" "$file" &&
-						rmdir "$GA_RC_TEMP_DIR"; then
-						echo TRANSFER-SUCCESS RETRIEVE "$key"
-					else
-						echo TRANSFER-FAILURE RETRIEVE "$key"
-					fi
-				;;
-			esac
-		;;
-		CHECKPRESENT)
-			key="$2"
-			calclocation "$key"
-
-			do_checkpresent "$key" "$LOC$key"
-		;;
-		REMOVE)
-			key="$2"
-			calclocation "$key"
-
-			do_remove "$key" "$LOC$key"
-		;;
-		*)
-			# The requests listed above are all the ones
-			# that are required to be supported, so it's fine
-			# to say that any other request is unsupported.
-			echo UNSUPPORTED-REQUEST
-		;;
-	esac
-done
-
 
-# XXX anything that needs to be done at shutdown can be done here
+import os
+import re
+import sys
+import shutil
+import tempfile
+import subprocess
+
+def say(cmd, *messages):
+    print(cmd, *messages)
+    sys.stdout.flush()
+
+def ask(cmd, *query):
+    say(cmd, *query)
+    resp_key, _, resp_value = sys.stdin.readline().partition(" ")
+    if resp_key == "VALUE":
+        return resp_value.rstrip()
+    if resp_key == "CREDS":
+        return resp_value.rstrip().split(" ", 1)
+
+def rclone(*args):
+    proc = subprocess.Popen(("rclone",) + args, stderr=subprocess.STDOUT, stdout=subprocess.PIPE)
+    output = proc.stdout.read()
+    return proc.wait(), output.decode("utf8")
+
+def get_config(key):
+    return ask("GETCONFIG", key)
+
+def set_config(key, value):
+    say("SETCONFIG", key, value)
+
+def debug(*messages):
+    say("DEBUG", " ".join(map(str, messages)))
+
+
+class Remote(object):
+
+    def initremote(self):
+        try:
+            self.remote_prefix = get_config("prefix")
+            if not self.remote_prefix:
+                self.remote_prefix = "git-annex"
+            if self.remote_prefix == "/":
+                raise RuntimeError("storing objects directly in the root (/) is not supported")
+            set_config("prefix", self.remote_prefix)
+
+            self.remote_target = get_config("target")
+            set_config("target", self.remote_target)
+
+            self.rclone_layout = get_config("rclone_layout") or "lower"
```

Comment:
Why isn't the `or "lower"` bit in `self._validate_layout()` like in the bash version?